Sparc: RzIL

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

Depends on:
- https://github.com/rizinorg/rizin/pull/5308 (basis)
- https://github.com/rizinorg/rizin/pull/5304
- https://github.com/rizinorg/rizin/pull/5305
- https://github.com/rizinorg/rizin/pull/5306

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

ISA: https://www.oracle.com/technetwork/server-storage/sun-sparc-enterprise/documentation/sparc-architecture-2015-2868130.pdf

Todo

- [x] Double/quad float loading addresses incorrect register.
- [x] LDX/A should have same semantics as LDW, LDH etc.
- [x] Remove else case of `if (n < 1) {}` in `analyze_op` to save indentations.
- [x] Replace `VARG` with macro to always return `0` for `g0` reg.
- [x] Update expected rz-tracetest results
- [x] FMUL8X16 should round for negative and positive numbers.
- [x] QEMU gdbstub misses all ASR registers from ASR7 upwards -.- (QEMU handles only GSR anyways)

Testing

- [x] All kind of branching tests
- [x] Stack save/restore
- [x] `ccr` reg testing
- [x] Floats
- [x] memory ld/st
- [x] Saturation
- [x] After validation. Add all Capstone MC instructions to ASM tests
- [x] ~~Little endian branch binary test~~ Doesn't seem to make sense. No little endian sensitive instructions there.
- [x] SIMD instructions

**Others**

`bn <addr>`/`fbn` saves the incorrect xref.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added

**QEMU instructions covered by rz-tracetest**

<img width="1084" height="105" alt="image" src="https://github.com/user-attachments/assets/88bac316-66dc-4893-9587-3af1ff107d93" />


**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

closes https://github.com/rizinorg/rizin/issues/5179

notxvilka commented on subprojects/capstone-dev.wrap:
```diff
@@ -0,0 +1,6 @@
+[wrap-git]
+url = git@github.com:Rot127/capstone.git
```
It will not work on the CI, please use `https://` link instead.

Rot127 replied:
Yes, this is still the old branch. Switch back soon.

notxvilka reviewed (changes requested):
1. Please rebase
2. Extract non-SPARC changes into a separate PR.

Rot127 committed:
Remove GNU Sparc asm module.

Rot127 committed:
Add macros for LOAD/STORE that allow to specify the memory address space.

Rot127 committed:
Uplift Sparc to RzIL.

This uplifts all non-trap related instructions to RzIL.
Notable shortcomings are:
- ASI (address spaces identifiers) are not implemented.
  Every instruction using them defaults to the ASI 0.
  This is because STORE/LOAD don't take RzIL values but constants for
  memory identifiers.
- Similary floating point rounding defaults always to one rounding mode.
  See SPARC_RMODE(). For the same reason as above, the rounding mode
  can't be determined during runtime currently.
- Floating point exception flags (FSR.cexc/FSR.aexc) are not written.
  This also happens because RzIL has currently no mechanism to hook FP
  exceptions from the FP module and execute predefined RzIL code for them.

notxvilka reviewed (changes requested):
```
In file included from ../librz/arch/isa/sparc/sparc_il.c:8:
../librz/arch/isa/sparc/sparc_il.c: In function ‘rz_sparc_2args_ccr’:
../librz/arch/isa/sparc/sparc_il.c:507:37: error: ‘il_op’ may be used uninitialized [-Werror=maybe-uninitialized]
  507 |         res128 = SETL("res128_mid", il_op(VARL("arg0"), VARL("arg1")));
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../librz/include/rz_il/rz_il_opbuilder_begin.h:163:57: note: in definition of macro ‘SETL’
  163 | #define SETL(name, v)      rz_il_op_new_set(name, true, v)
      |                                                         ^
../librz/arch/isa/sparc/sparc_il.c:452:30: note: ‘il_op’ was declared here
  452 |         rz_il_pure_2args_op *il_op;
      |                              ^~~~~
cc1: all warnings being treated as errors
```

notxvilka commented:
Unrelated to the SPARC directly but came from updating Capstone, I guess:
```
../librz/arch/p/analysis/analysis_loongarch_cs.c:157:14: runtime error: index -1 out of bounds for type 'RzAnalysisValue *[8]' (aka 'struct rz_analysis_value_t *[8]')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../librz/arch/p/analysis/analysis_loongarch_cs.c:157:14 

../librz/arch/p/analysis/analysis_loongarch_cs.c:687:49: runtime error: left shift of negative value -203
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../librz/arch/p/analysis/analysis_loongarch_cs.c:687:49 
../librz/arch/p/analysis/analysis_loongarch_cs.c:697:49: runtime error: left shift of negative value -524288
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../librz/arch/p/analysis/analysis_loongarch_cs.c:697:49 
../librz/arch/p/analysis/analysis_loongarch_cs.c:692:49: runtime error: left shift of negative value -524288
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../librz/arch/p/analysis/analysis_loongarch_cs.c:692:49 
ERROR: Linear size differs too much from the bbsum, please use pdr instead.
ERROR: core: Invalid address (fcn.00010208)
ERROR: core: Error while executing command: pdf @ fcn.00010208
```

notxvilka reviewed (changes requested):
Update also tests in `db/cmd/cmd_list` that list plugins.

Rot127 commented on librz/bin/format/elf/elf_relocs_patching.c:
```diff
@@ -1047,13 +1326,15 @@ void Elf_(rz_bin_elf_patch_relocation)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL RzBinE
 		.P = rel->vaddr,
 		.MB = 0,
-		.G = 0,
+		.G = Elf_(rz_bin_get_reloc_sym_offset_in_got)(bin, rel->sym),
```
@notxvilka @wargio Is this an acceptable runtime overhead? It adds something like `O(n^2): n = number of relocs`.
I have no feeling about how many relocs there are usually.
Otherwise, I would just add a check to return early.

notxvilka replied:
Usually there are less than hundred per binary, rarely thousands. So it should be acceptable for now. We can think on how to improve in the separate PR to address some relocations shortcomings.

wargio replied:
i agree. should be ok

notxvilka commented on librz/bin/format/elf/elf.h:
```diff
@@ -147,17 +147,28 @@ typedef struct rz_bin_elf_symbol_t {
+	/**
+	 * \brief Architecture specific ELF relocation values.
+	 */
+	union {
```
@wargio @thestr4ng3r @kazarmy what is your opinion on that?

wargio replied:
with just one element makes no sense.

kazarmy replied:
i think what is meant here is that the union either exists or it doesn't, but this is not C++, so misuse of the enclosed var probably is best detected by assertion rather than enforced syntactically

Rot127 replied:
I put it into a union out of reflex. Expecting other archs might have special values encoded in `r_info` as well. But will remove the union and make it a normal member.

notxvilka commented on librz/bin/format/elf/elf_relocs_patching.c:
```diff
@@ -13,11 +13,17 @@ typedef struct reloc_formular_symbols_t {
+	 * \brief Sparc64: Second addend, extracted from the r_info field by
+	 * applying the ELF_SPARC64_R_TYPE_DATA macro (alias for ELF64_R_TYPE_DATA).
+	 */
+	ut64 O;
```
@wargio @thestr4ng3r @kazarmy and here

kazarmy replied:
a possible source reference appears to be https://docs.oracle.com/en/operating-systems/solaris/oracle-solaris/11.4/linkers-libraries/64-bit-sparc-relocation-types.html ... looks reasonable enough

kazarmy replied:
i would be strict in putting "secondary" instead of "second" though, since constant + constant is yet another constant

notxvilka commented on librz/util/bits.c:
```diff
@@ -0,0 +1,36 @@
+ * \param value The value to spread.
+ * \return The result of setting the \p value bits to the bit positions of \p mask.
+ */
+RZ_API ut64 rz_bits_spread(const ut64 mask, const ut64 value) {
```
Extract this into a separate PR please

notxvilka commented on librz/il/theory_bitv.c:
```diff
@@ -230,6 +230,7 @@ void *rz_il_handler_div(RzILVM *vm, RzILOpBitVector *op, RzILTypePure *type) {
 		if (rz_bv_is_zero_vector(y)) {
```
Extract all non-SPARC RzIL changes (including a small PIC modification) into a separate PR that we will merge before

notxvilka commented on test/unit/test_bits.c:
```diff
@@ -35,8 +35,23 @@ bool test_rz_bits_count(void) {
 	mu_end;
 }
 
+bool test_rz_bits_spread(void) {
```
Same, together with `librz/util/bits.c` into a separate PR

notxvilka reviewed (changes requested):
Just a request to untangle some of the changes

Rot127 committed:
Implement Sparc relocation patching and conversion.

notxvilka commented:
@Rot127 please rebase again (and also update Capstone commit one more time) since corresponding PRs were merged.

notxvilka reviewed (changes requested):
On Windows tests produce different output:
```
[XX]   1 secs C:\projects\rizin\test\db\rzil\sparc Run jump instructions - Sparc32
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.9.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -eio.cache=1 -qc 's 0x000106b8
aezi
aezsu 0x000106b0
aezvt | sort
' bins/elf/sparc/sparc32_insn_jmp.bin
-- stdout
--- expected
+++ actual
@@ -0,7 +0,6 @@
 
 ---------------------------
-PC         bitv 0x000106b0
 asi        bitv 0x00000000
 asr        bitv 0x00000000
 asr0       bitv 0x00000000
@@ -108,6 +107,7 @@
 o5         bitv 0x00000000
 o7         bitv 0x00000000
 otherwin   bitv 0x00000000
+PC         bitv 0x000106b0
 pil        bitv 0x00000000
 psr        bitv 0x00000000
 pstate     bitv 0x00000000
[XX]   1 secs C:\projects\rizin\test\db\rzil\sparc Run jump instructions - Sparc64
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.9.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -eio.cache=1 -qc 's 0x00101078
aezi
aezsu 0x0010106c
aezvt | sort
' bins/elf/sparc/sparc64_insn_jmp.bin
-- stdout
--- expected
+++ actual
@@ -0,7 +0,6 @@
 
 -----------------------------------
-PC         bitv 0x000000000010106c
 asi        bitv 0x0000000000000000
 asr        bitv 0x0000000000000000
 asr0       bitv 0x0000000000000000
@@ -125,6 +124,7 @@
 o5         bitv 0x0000000000000000
 o7         bitv 0x0000000000000000
 otherwin   bitv 0x0000000000000000
+PC         bitv 0x000000000010106c
 pil        bitv 0x0000000000000000
 pstate     bitv 0x0000000000000000
 sp         bitv 0x0000000000000000
[XX]   1 secs C:\projects\rizin\test\db\rzil\sparc Run all instructions - Sparc64
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.9.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -eio.cache=1 -qc 's 0x00101ee0
aezi
aezsu 0x00101ed8
aezvt | sort
' bins/elf/sparc/sparc64_insn_all.bin
-- stdout
--- expected
+++ actual
@@ -0,7 +0,6 @@
 
 -----------------------------------
-PC         bitv 0x0000000000101ed8
 asi        bitv 0x0000000000000000
 asr        bitv 0x0000000000000000
 asr0       bitv 0x0000000000000000
@@ -125,6 +124,7 @@
 o5         bitv 0x0000000000000000
 o7         bitv 0x0000000000000000
 otherwin   bitv 0x0000000000000000
+PC         bitv 0x0000000000101ed8
 pil        bitv 0x0000000000000000
 pstate     bitv 0x0000000000000000
 sp         bitv 0x0000000000000000
[XX]   1 secs C:\projects\rizin\test\db\rzil\sparc Run all instructions - Sparc32
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.9.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -eio.cache=1 -qc 's 0x00011690
aezi
aezsu 0x00010464
aezvt | sort
' bins/elf/sparc/sparc32_insn_all.bin
-- stdout
--- expected
+++ actual
@@ -0,7 +0,6 @@
 
 ---------------------------
-PC         bitv 0x00010464
 asi        bitv 0x00000000
 asr        bitv 0x00000000
 asr0       bitv 0x00000000
@@ -108,6 +107,7 @@
 o5         bitv 0xdddddddd
 o7         bitv 0x00000000
 otherwin   bitv 0x00000000
+PC         bitv 0x00010464
 pil        bitv 0x00000000
 psr        bitv 0x00000000
 pstate     bitv 0x00000000
```
See https://ci.appveyor.com/project/rizinorg/rizin/builds/52604006/job/k322uv2gfvaef5w9

notxvilka reviewed (changes requested):
Some tests are broken. Also:
```
WARNING: Unhandled conversion of ELF reloc 21 (R_SPARC_JMP_SLOT)
WARNING: Unhandled conversion of ELF reloc 19 (R_SPARC_COPY)
```

Rot127 commented:
The Windows job looks weird. All tests passed but it is complaining about a build time of 90min+

notxvilka commented:
> The Windows job looks weird. All tests passed but it is complaining about a build time of 90min+

Because it has a timeout limit of 1h30 minutes. For now we can restart, but we need to think about optimizing some tests.

wargio commented:
It's probably due the swift c++ code
